### DISCUSSION TITLE
some functions do not work in Larch 0.9.70

### DISCUSSION THREAD

**Discussion by zhangnaichi2020** [xraypy/xraylarch#450] [Q&A] [answered]
I'm a beginner on how to use larch code , and I want to run larch with python in jupyter, so I import the functions that will be used in a way like 
"
import larch
from larch import Group
from larch.builtins import _journal
from larch.io import read_ascii
from larch.xafs import sort_xafs, pre_edge
from larch.math.utils import boxcar, savitzky_golay
from larch.xafs.deconvolve import xas_convolve, xas_deconvolve
"
Here is my questions:
1. after update the larch version from 0.9.68 (maybe) to the latest 0.9.70, it takes much longer to run the pre_edge function;
2. for xas_deconvolve and xas_convolve
    1) xas_convolve runs much much slower than xas_deconvolve
    2) for both function, if i run 
"
dat = read_ascii('./Cr_standards/CrOH3_Cr_K.xdi')
xas_deconvolve(dat, esigma=1.0)  # or xas_convolve(dat, esigma=1.0)
"
it returns "Warning: xas_deconvolve: needs First Argument Group or valid arguments for energy, norm", and it goes well when i run
 "xas_deconvolve(dat.energy, dat.mu, esigma=1.0)"
it seems does not follow the First Argument Group convention;
    3) when i run 
"
dat = read_ascii('./Cr_standards/CrOH3_Cr_K.xdi')
xas_deconvolve(dat, esigma=1.0) 
dat.deconv
"
it returns "AttributeError: 'Group' object has no attribute 'deconv'", which is not conform with the samples given in 13.2.7.3.

**Comment by newville** [comment, marked as answer]
@zhangnaichi2020 I think this because `xas_deconvolve` expects to work on normalized XAFS data, that is after running `pre_edge()`.



```
 dat = read_ascii('./Cr_standards/CrOH3_Cr_K.xdi')
pre_edge(dat)
xas_deconvolve(dat, esigma=1.0) 
```

because otherwise it is just not clear what to do the deconvolution on

And,  yes, we did see that `pre_edge()` could get very slow for some input data (especially, with repeated energy values) with changes in 0.9.70.   That should be fixed in the master branch, which you can install with

```
  python -m pip install https://millenia.cars.aps.anl.gov/xraylarch/downloads/xraylarch-latest-py3-none-any.whl
```
